src/core/lib/resource_quota/memory_quota.cc

```diff
@@ -0,0 +1,375 @@
+// Copyright 2021 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <grpc/support/port_platform.h>
+
+#include "src/core/lib/resource_quota/memory_quota.h"
+
+#include <thread>
+
+#include "src/core/lib/promise/loop.h"
+#include "src/core/lib/promise/race.h"
+#include "src/core/lib/promise/seq.h"
+
+namespace grpc_core {
+
+//
+// Reclaimer
+//
+
+ReclamationSweep::~ReclamationSweep() {
+  if (memory_quota_ != nullptr) {
+    memory_quota_->FinishReclamation(sweep_token_);
+  }
+}
+
+//
+// MemoryRequest
+//
+
+namespace {
+size_t RoundUp(size_t size, size_t block_size) {
+  return (size + block_size - 1) / block_size * block_size;
+}
+size_t RoundDown(size_t size, size_t block_size) {
+  return size / block_size * block_size;
+}
+}  // namespace
+
+MemoryRequest MemoryRequest::WithBlockSize(size_t block_size) const {
+  MemoryRequest r(*this);
+  r.block_size_ = block_size;
```

Comment:
Can we make block_size a ctor arg for `MemoryRequest`?

Reply:
We could... these helpers should be inline anyway (which makes the perf about moot) -- I like that this way of writing it out calls out that you're doing something funky -- majority of callers shouldn't care about block size.

Reply:
removing block size for this round -- it's probably unlikely to be actually used.